### PR TITLE
Define and emit filter metadata for AOT (Issue #664)

### PR DESCRIPTION
**Pull request by steven-johnson**
Not sure if this is ready to land as-is (look for some TODO issues),
but take a look-see to see if this looks like the right direction.

### CODE REVIEW

**Comment by steven-johnson** [comment]
(Also note that this goes all-in for C++11 in mainline code; I even used lambdas. Speak up now if this disturbs you.)

File: src/CodeGen_LLVM.cpp [hunk 1]
```diff
@@ -602,6 +618,113 @@ void CodeGen_LLVM::compile_buffer(const Buffer &buf) {
...
+    if (!e.defined() || e.type().code == Type::Handle) {
+        // Handle is always emitted as "undefined"
+        return Constant::getNullValue(scalar_value_t_type->getPointerTo());
+    }
+
+    // This function is typically called exactly once (at most) per
+    // instance of CodeGen_LLVM, so it's not really worth trying to hoist
+    // out of this function. (It also uses the llvm::Type member variables
+    // from CodeGen_LLVM; keeping it here makes it easier to just re-use those.)
+    std::map<int, std::function<llvm::Constant *(Expr e)>> expr_to_llvm_constant_func_map{
```
Comment (abadams) [inline src/CodeGen_LLVM.cpp]: Creating and using this map seems more verbose than the equivalent if tree:

llvm::Constant *constant = NULL;
if (e.type() == Bool()) {
  constant = get_constant<bool>(i1, e);
} else if (e.type() == Float(...)) {
  constant = ConstantFP::get(...);
} else if (e.type() == Int(...)) {
  constant = ConstantInt::get(...);
} else if (...) {
  ...
} else {
  internal_error << "Unhandled...";
}

File: src/runtime/HalideRuntime.h [hunk 2]
```diff
@@ -354,6 +354,75 @@ typedef struct buffer_t {
...
+  uint16_t u16;
+  uint32_t u32;
+  uint64_t u64;
+  float f32;
+  double f64;
+  void *handle;
+};
+
+enum halide_argument_kind_t {
+    HalideArgumentKind_InputScalar = 0,
```
Comment (abadams) [inline src/runtime/HalideRuntime.h]: The other enums in HalideRuntime.h look_like_this

File: src/runtime/HalideRuntime.h [hunk 3]
```diff
@@ -354,6 +354,75 @@ typedef struct buffer_t {
 
 #endif
 
+/** halide_scalar_value_t is a simple union able to represent all the well-known
+ * scalar values in a filter argument. Note that it isn't tagged with a type;
+ * you must ensure you know the proper type before accessing. Most user
+ * code will never need to create instances of this struct; its primary use
+ * is to hold def/min/max values in a halide_filter_argument_t. */
+union halide_scalar_value_t {
+  bool b;
```
Comment (abadams) [inline src/runtime/HalideRuntime.h]: four-space indentation please

File: src/runtime/HalideRuntime.h [hunk 4]
```diff
@@ -354,6 +354,75 @@ typedef struct buffer_t {
...
+  // and *may* be null for scalar arguments. (A null value means )
+  const halide_scalar_value_t *def;
+  const halide_scalar_value_t *min;
+  const halide_scalar_value_t *max;
+};
+#pragma pack(pop)
+
+#pragma pack(push, 1)
+struct halide_filter_metadata_t {
+  // TODO: should we add a version field at the beginning, in case we want
```
Comment (abadams) [inline src/runtime/HalideRuntime.h]: I'd vote for a version field. It's easy to add, and easy to regret not adding one.

File: src/runtime/HalideRuntime.h [hunk 5]
```diff
@@ -354,6 +354,75 @@ typedef struct buffer_t {
...
+#pragma pack(push, 1)
+struct halide_filter_argument_t {
+  const char *name;       // name of the argument; will never be null or empty.
+  // TODO: all these would fit into uint8. Should we pack them tightly?
+  int32_t kind;           // actually halide_argument_kind_t
+  int32_t dimensions;     // always zero for scalar arguments
+  int32_t type_code;      // actually halide_type_code_t
+  int32_t type_bits;      // [1, 8, 16, 32, 64]
+  // These pointers should always be null for buffer arguments,
+  // and *may* be null for scalar arguments. (A null value means )
```
Comment (abadams) [inline src/runtime/HalideRuntime.h]: Don't forget to finish the comment about what a null value means.

Comment (steven-johnson) [inline src/runtime/HalideRuntime.h]: I dunno, I kinda like the stark nihilism of the existing form

File: src/CodeGen_LLVM.cpp (hunk 1 above)
Comment (steven-johnson) [inline src/CodeGen_LLVM.cpp]: Hmm, you could be right (though we need a lot more else-if's than your example above). I'll experiment and see.

**Comment by steven-johnson** [comment]
PTAL

File: test/generator/metadata_tester_aottest.cpp [hunk 6]
```diff
@@ -0,0 +1,397 @@
+#include "HalideRuntime.h"
```
Comment (abadams) [inline test/generator/metadata_tester_aottest.cpp]: More two-space indentation in this file

**Comment by abadams** [comment]
lgtm except for two remaining nits

File: test/generator/metadata_tester_aottest.cpp (hunk 6 above)
Comment (steven-johnson) [inline test/generator/metadata_tester_aottest.cpp]: I have got to fix my editor to adapt to 2-vs-4 automatically

Comment (steven-johnson) [inline test/generator/metadata_tester_aottest.cpp]: Ran this file through clang-format with our default settings; hopefully it's OK now.

**Comment by steven-johnson** [comment]
PTAL

**Pull request merged**